// jet-validator/lib/jet-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERR_MSG = void 0;
exports.ERR_MSG = 'The following parameter was missing or invalid: ';
function jetValidator(errCode) {
    return (...params) => {
        const itemArr = getLoopFns(params);
        return (req, res, next) => {
            for (const item of itemArr) {
                if (!item.fn(req)) {
                    const error = exports.ERR_MSG + `"${item.paramName}".`;
                    return res.status(errCode !== null && errCode !== void 0 ? errCode : 400).json({ error });
                }
            }
            return next();
        };
    };
}
function getLoopFns(params) {
    const retVal = [];
    for (const param of params) {
        const { paramName, type, reqObjProp } = getParamFields(param);
        let loopFn = () => false;
        if (type === 'string') {
            loopFn = checkStr(reqObjProp, paramName);
        }
        else if (type === 'number') {
            loopFn = checkNum(reqObjProp, paramName);
        }
        else if (type === 'boolean') {
            loopFn = checkBool(reqObjProp, paramName);
        }
        else if (type === 'function') {
            const fn = param[1];
            loopFn = checkValidatorFn(reqObjProp, paramName, fn);
        }
        else {
            loopFn = checkDefault(reqObjProp, paramName, type);
        }
        retVal.push({
            paramName,
            fn: loopFn,
        });
    }
    return retVal;
}
function getParamFields(param) {
    let paramName = '';
    let type = 'string';
    let reqObjProp = 'body';
    if (typeof param === 'string') {
        paramName = param;
    }
    else if (param instanceof Array) {
        paramName = param[0];
        if (typeof paramName !== 'string') {
            throw Error('param name must be a string');
        }
        if (param.length > 1) {
            if (typeof param[1] === 'string') {
                type = param[1];
            }
            else if (typeof param[1] === 'function') {
                type = 'function';
            }
            else {
                throw Error('param[1] must be a string or a validator function');
            }
        }
        const prop = (param.length >= 3 ? param[2] : 'body');
        if (prop !== 'body' && prop !== 'params' && prop !== 'query') {
            throw Error('param[2] must be "body", "query", or "params"');
        }
        else {
            reqObjProp = prop;
        }
    }
    else {
        throw Error('"validate()" argument must be a string or array');
    }
    return { paramName, type, reqObjProp };
}
function checkStr(reqObjProp, paramName) {
    return getLoopFn(reqObjProp, paramName, (toCheck) => {
        return (typeof toCheck === 'string');
    });
}
function checkNum(reqObjProp, paramName) {
    return getLoopFn(reqObjProp, paramName, (toCheck) => {
        if (reqObjProp === 'query' || reqObjProp === 'params') {
            return !isNaN(toCheck);
        }
        else if (reqObjProp === 'body' && typeof toCheck !== 'number') {
            return false;
        }
        else {
            return true;
        }
    });
}
function checkBool(reqObjProp, paramName) {
    return getLoopFn(reqObjProp, paramName, (toCheck) => {
        if (reqObjProp === 'query' || reqObjProp === 'params') {
            return isBool(toCheck);
        }
        else if (reqObjProp === 'body' && typeof toCheck !== 'boolean') {
            return false;
        }
        else {
            return true;
        }
    });
}
function isBool(arg) {
    if (typeof arg === 'boolean') {
        return true;
    }
    else if (typeof arg === 'string') {
        arg = arg.toLowerCase();
        return (arg === 'true' || arg === 'false');
    }
    else {
        return false;
    }
}
function checkValidatorFn(reqObjProp, paramName, validatorFn) {
    return getLoopFn(reqObjProp, paramName, (toCheck) => {
        return validatorFn(toCheck);
    });
}
function checkDefault(reqObjProp, paramName, type) {
    return getLoopFn(reqObjProp, paramName, (toCheck) => {
        return (typeof toCheck === type);
    });
}
function getLoopFn(reqObjProp, paramName, cb) {
    return (req) => {
        const toCheck = req[reqObjProp][paramName];
        return cb(toCheck);
    };
}
exports.default = jetValidator;
